utils/filesystem.c

```diff
@@ -0,0 +1,414 @@
+#include "filesystem.h"
+#include "tsh.h"
+
+#include <dirent.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <libgen.h>
+#include <limits.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/xattr.h>
+#include <unistd.h>
+
+static void close_fd(int *fd)
+{
+    if (!fd || *fd < 0)
+        return;
+
+    close(*fd);
+    *fd = -1;
+}
+
+bool pv_filesystem_path_exist(const char *path)
+{
+    return access(path, F_OK) == 0;
+}
+
+bool pv_filesystem_path_is_directory(const char *path)
```

Comment:
Ok, this is a problem I had when I thought about this function and `pv_filesystem_path_exist`. The thing is, `pv_filesystempath_is_directory` checks if the current path is a directory, so return `false` if the path refers to a file. If I change the name to `pv_filesystem_dir_exist`, you could try to use the function to check if it's necessary to create a folder, something like this:
```C
if (!pv_filesystem_dir_exist(path))
    pv_filesystem_mkdir_p(path);
```
So, if `path` is a file, the code has a bug because we cannot create a folder and file with the same name.
That's the reason to create these two functions